File: clients/python/src/aod/resources/agents.py
from __future__ import annotations

from typing import Any
from uuid import UUID

import httpx
from pydantic import BaseModel

from .._http import check_response
from ..models import Agent, AgentVersion, SkillInput


def _normalize_skills(skills: list[SkillInput] | None) -> list[dict[str, Any]] | None:
    """Accept typed pydantic inputs or raw dicts, return wire form."""
    if skills is None:
        return None
    out: list[dict[str, Any]] = []
    for entry in skills:
        if isinstance(entry, BaseModel):
            out.append(entry.model_dump(exclude_none=True))
        else:
            out.append(entry)
    return out


def _create_body(
    *,
    name: str,
    model: str,
    runtime: str,
    system: str | None = None,
    description: str | None = None,
    environment_id: str | UUID | None = None,
    skills: list[SkillInput] | None = None,
    mcp_servers: list[dict[str, Any]] | None = None,
    metadata: dict[str, Any] | None = None,
) -> dict[str, Any]:
    body: dict[str, Any] = {"name": name, "model": model, "runtime": runtime}
    if system is not None:
        body["system"] = system
    if description is not None:
        body["description"] = description
    if environment_id is not None:
        body["environment_id"] = str(environment_id)
    normalized_skills = _normalize_skills(skills)
    if normalized_skills is not None:
        body["skills"] = normalized_skills
    if mcp_servers is not None:
        body["mcp_servers"] = mcp_servers
    if metadata is not None:
        body["metadata"] = metadata
    return body


def _update_body(
    *,
    version: int,
    name: str | None = None,
    model: str | None = None,
    runtime: str | None = None,
    system: str | None = None,
    description: str | None = None,
    environment_id: str | UUID | None = None,
    skills: list[SkillInput] | None = None,
    mcp_servers: list[dict[str, Any]] | None = None,
    metadata: dict[str, Any] | None = None,
) -> dict[str, Any]:
    body: dict[str, Any] = {"version": version}
    if name is not None:
        body["name"] = name
    if model is not None:
        body["model"] = model
    if runtime is not None:
        body["runtime"] = runtime
    if system is not None:
        body["system"] = system
    if description is not None:
        body["description"] = description
    if environment_id is not None:
        body["environment_id"] = str(environment_id)
    normalized_skills = _normalize_skills(skills)
    if normalized_skills is not None:
        body["skills"] = normalized_skills
    if mcp_servers is not None:
        body["mcp_servers"] = mcp_servers
    if metadata is not None:
        body["metadata"] = metadata
    return body


class Agents:
    def __init__(self, client: httpx.Client) -> None:
        self._client = client

    def list(self) -> list[Agent]:
        body = check_response(self._client.get("/agents"))
        return [Agent.model_validate(a) for a in body["data"]]

    def create(
        self,
        *,
        name: str,
        model: str,
        runtime: str,
        system: str | None = None,
        description: str | None = None,
        environment_id: str | UUID | None = None,
        skills: list[SkillInput] | None = None,
        mcp_servers: list[dict[str, Any]] | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> Agent:
        body = _create_body(
            name=name,
            model=model,
            runtime=runtime,
            system=system,
            description=description,
            environment_id=environment_id,
            skills=skills,
            mcp_servers=mcp_servers,
            metadata=metadata,
        )
        return Agent.model_validate(check_response(self._client.post("/agents", json=body)))

    def get(self, agent_id: str | UUID) -> Agent:
        return Agent.model_validate(check_response(self._client.get(f"/agents/{agent_id}")))

    def update(
        self,
        agent_id: str | UUID,
        *,
        version: int,
        name: str | None = None,
        model: str | None = None,
        runtime: str | None = None,
        system: str | None = None,
        description: str | None = None,
        environment_id: str | UUID | None = None,
        skills: list[SkillInput] | None = None,
        mcp_servers: list[dict[str, Any]] | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> Agent:
        body = _update_body(
            version=version,
            name=name,
            model=model,
            runtime=runtime,
            system=system,
            description=description,
            environment_id=environment_id,
            skills=skills,
            mcp_servers=mcp_servers,
            metadata=metadata,
        )
        return Agent.model_validate(
            check_response(self._client.put(f"/agents/{agent_id}", json=body))
        )

    def archive(self, agent_id: str | UUID) -> Agent:
        return Agent.model_validate(
            check_response(self._client.post(f"/agents/{agent_id}/archive"))
        )

    def versions(self, agent_id: str | UUID) -> list[AgentVersion]:
        body = check_response(self._client.get(f"/agents/{agent_id}/versions"))
        return [AgentVersion.model_validate(v) for v in body["data"]]


class AsyncAgents:
    def __init__(self, client: httpx.AsyncClient) -> None:
        self._client = client

    async def list(self) -> list[Agent]:
        body = check_response(await self._client.get("/agents"))
        return [Agent.model_validate(a) for a in body["data"]]

    async def create(
        self,
        *,
        name: str,
        model: str,
        runtime: str,
        system: str | None = None,
        description: str | None = None,
        environment_id: str | UUID | None = None,
        skills: list[SkillInput] | None = None,
        mcp_servers: list[dict[str, Any]] | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> Agent:
        body = _create_body(
            name=name,
            model=model,
            runtime=runtime,
            system=system,
            description=description,
            environment_id=environment_id,
            skills=skills,
            mcp_servers=mcp_servers,
            metadata=metadata,
        )
        return Agent.model_validate(check_response(await self._client.post("/agents", json=body)))

    async def get(self, agent_id: str | UUID) -> Agent:
        return Agent.model_validate(check_response(await self._client.get(f"/agents/{agent_id}")))

    async def update(
        self,
        agent_id: str | UUID,
        *,
        version: int,
        name: str | None = None,
        model: str | None = None,
        runtime: str | None = None,
        system: str | None = None,
        description: str | None = None,
        environment_id: str | UUID | None = None,
        skills: list[SkillInput] | None = None,
        mcp_servers: list[dict[str, Any]] | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> Agent:
        body = _update_body(
            version=version,
            name=name,
            model=model,
            runtime=runtime,
            system=system,
            description=description,
            environment_id=environment_id,
            skills=skills,
            mcp_servers=mcp_servers,
            metadata=metadata,
        )
        return Agent.model_validate(
            check_response(await self._client.put(f"/agents/{agent_id}", json=body))
        )

    async def archive(self, agent_id: str | UUID) -> Agent:
        return Agent.model_validate(
            check_response(await self._client.post(f"/agents/{agent_id}/archive"))
        )

    async def versions(self, agent_id: str | UUID) -> list[AgentVersion]:
        body = check_response(await self._client.get(f"/agents/{agent_id}/versions"))
        return [AgentVersion.model_validate(v) for v in body["data"]]

File: clients/typescript/src/resources/agents.ts
import type { HttpClient } from "../http.js";
import type { Agent, AgentVersion, McpServer } from "../types.js";

export interface AgentCreateParams {
  name: string;
  model: string;
  runtime: string;
  system?: string;
  description?: string;
  environment_id?: string;
  skills?: Record<string, unknown>[];
  mcp_servers?: McpServer[];
  metadata?: Record<string, unknown>;
}

export interface AgentUpdateParams {
  version: number;
  name?: string;
  model?: string;
  runtime?: string;
  system?: string;
  description?: string;
  // Pass `null` to detach the current environment from the agent.
  environment_id?: string | null;
  skills?: Record<string, unknown>[];
  mcp_servers?: McpServer[];
  metadata?: Record<string, unknown>;
}

interface ListResponse<T> {
  data: T[];
}

export class Agents {
  constructor(private readonly http: HttpClient) {}

  async list(opts: { signal?: AbortSignal } = {}): Promise<Agent[]> {
    const body = await this.http.request<ListResponse<Agent>>("GET", "/agents", {
      signal: opts.signal,
    });
    return body.data;
  }

  async create(
    params: AgentCreateParams,
    opts: { signal?: AbortSignal } = {},
  ): Promise<Agent> {
    return this.http.request<Agent>("POST", "/agents", {
      body: stripUndefined(params),
      signal: opts.signal,
    });
  }

  async get(agentId: string, opts: { signal?: AbortSignal } = {}): Promise<Agent> {
    return this.http.request<Agent>("GET", `/agents/${agentId}`, {
      signal: opts.signal,
    });
  }

  async update(
    agentId: string,
    params: AgentUpdateParams,
    opts: { signal?: AbortSignal } = {},
  ): Promise<Agent> {
    return this.http.request<Agent>("PUT", `/agents/${agentId}`, {
      body: stripUndefined(params),
      signal: opts.signal,
    });
  }

  async archive(agentId: string, opts: { signal?: AbortSignal } = {}): Promise<Agent> {
    return this.http.request<Agent>("POST", `/agents/${agentId}/archive`, {
      signal: opts.signal,
    });
  }

  async versions(
    agentId: string,
    opts: { signal?: AbortSignal } = {},
  ): Promise<AgentVersion[]> {
    const body = await this.http.request<ListResponse<AgentVersion>>(
      "GET",
      `/agents/${agentId}/versions`,
      { signal: opts.signal },
    );
    return body.data;
  }
}

function stripUndefined<T extends object>(obj: T): Partial<T> {
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(obj)) {
    if (v !== undefined) out[k] = v;
  }
  return out as Partial<T>;
}
